### ISSUE TITLE
Format a Temporal object into a user-supplied string format

### ISSUE THREAD

**Issue by ptomato** [js-temporal/proposal-temporal-v2#5]
Similar to #2, but for string formatting instead of string parsing.

If #2 is adopted, an API using the same microformat could be proposed.

### Advantages:

Sometimes it's necessary to obtain a string that fits a particular format that is neither ISO 8601 nor locale-sensitive. For these cases, third-party date libraries usually provide a formatting API.

### Concerns:

In general, for human-readable strings, developers should not customize the format themselves, and use only the locale-specific formats provided by Intl.DateTimeFormat. While this functionality might be useful for formatting strings that must fit a certain format for machine-readability, providing such an API might result in worse user experiences because developers are tempted to use it for human-readable strings instead of Intl.DateTimeFormat.

Unlike parsing, this kind of functionality is easy to build in business logic.

### Prior art:

TBD

### Constraints / corner cases:

TBD

**Comment by CarterLi** [comment]
All I want is something like https://date-fns.org/v2.23.0/docs/format

eg. format(Date.now(), 'yyyy-MM-dd HH:mm:ss')

I have read [the doc of DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat) and I don't think it can match my requirement.

**Comment by ryzokuken** [comment]
I think strongly that Temporal should stick to the standards here. Format-transforms are a useful use-case for third-party modules.

**Comment by CarterLi** [comment]
> I think strongly that Temporal should stick to the standards here. Format-transforms are a useful use-case for third-party modules.

It IS the standard: https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table

3rd-party libraries like moment, dayjs, date-fns are already very easy to use and their APIs are familiar. People choose to a new standard object that is not very stable, has learning costs and browser compatibility concerns mainly because it can reduce 3rd party library dependencies and publication code sizes.

If people still have to use 3rd-party libs for such a common use case formatting string, why do they have to learn something completely different?

**Comment by ryzokuken** [comment]
@CarterLi I disagree. First of all, LDML is not a timestamp format standard, it is meant for localized formatting for i18n. We already allow localized formatting for Temporal objects via `Intl.DateTimeFormat`, so any such formatting can (and should) only be done using `Intl.DateTimeFormat` via `Temporal.Foo#toLocaleString`.

To a large extent, `Intl.DateTimeFormat` already allows customized formats like these, with the added benefit of making them locale-friendly.

**Comment by chrisveness** [comment]
I have a requirement to format some dates as either ‘8 Sep’ or ‘8 Sep 2020‘ (depending on whether the year is the current year or not).

`toLocaleString('en')` appears to insist that I want the month before the day, and the year separated by a comma – but I don't!

Is there a way to achieve those formats with the current APIs available?

If not, current APIs seem too rigid, and I would like to see custom formatting using tokens.

It has been an issue in the past that there have only been de-facto standards for such tokens, and it would be problematic to adopt any of these on a presumably arbitrary basis.

However, the Unicode Consortium _is_ a standards body, and as part of the Locale Data Markup Language (LDML) have defined a standard set of pattern fields for date formatting (https://unicode-org.github.io/cldr/ldml/tr35-dates.html#Date_Format_Patterns).

As LDML is intended explicitly for locale-specific applications, and includes references to non-Gregorian calendars (such as the Chinese lunar calendar), it looks as if it would make a good fit for the Temporal proposal.

I hope I don‘t migrate to Temporal, and then still have to load the whole of a third-party module such as _moment_ just to be able to generate custom formats.

Surely Unicode TR35 Part 4 (Dates) Section 8 (Date Format Patterns) would make a good basis  for Temporal custom formatting?

**Comment by ptomato** [comment]
The discussion thread starting from https://github.com/tc39/ecma402/issues/554#issuecomment-796903218 might prove illuminating as to why there is caution about a formatter API.

I'm not sure whether your requirement

> `toLocaleString('en')` appears to insist that I want the month before the day, and the year separated by a comma – but I don't!

means that you want to format dates with the day before the month and no comma, even in the `en` locale where that would be "incorrect" (at least, according to CLDR); or it means that you do actually want an internationalized date format but you are looking for the `en-GB` locale (or another `en-??` locale, but I guessed GB based on the location in your GitHub profile).

In any case `Temporal.Now.plainDateISO().toLocaleString('en-GB', { dateStyle: 'medium' })` seems to work for your purpose?

> I hope I don‘t migrate to Temporal, and then still have to load the whole of a third-party module such as _moment_ just to be able to generate custom formats.

It's more likely that a third-party module specifically for custom formatting would spring up, rather than loading Moment.

**Comment by craigmiller160** [comment]
I can't believe this is even a debate. Formatting strings based on user-supplied specifiers is a standard across a wide range of languages and libraries. Sticking to only the "standards" is the kind of decision that gets made in ivory tower rooms of computer theory and completely ignores the needs of programmers dealing with the real, messy world that we have to code for.

**Comment by ptomato** [comment]
I mean... this _is_ a repo for incubating a proposal for the standard, and there are other programmers in this messy world that have needs that would conflict with just copying the API from date-fns ... so I guess I don't understand why it's so unbelievable that there is a discussion about this? In any case, it would be helpful if you could provide some details about your use case. The more we know about how people would use this API, the better.

**Comment by Ariane-B** [comment]
Today, I wanted to get the following date/time format:

**French: `2022-12-30 à 20 h 22` OR  `2022-12-30 à 20:22`**

(20:22 isn't technically the format for use in sentences in French, but it's widely accepted as a more compact, more practical format to use for displaying times, compared to 20 h 22. Our design team isn't 100 % sure which format they want at this point.)

**English: `2022-12-30 at 20:22`**

I couldn't find a way to do that with the current `toLocaleString` or `DateTimeFormat` options on offer.

Maybe there's an obscure locale that allows it, but the lack of a full documentation of all supported locales and the formats they provide anywhere that I could find isn't making the search easy.

Anyway, with this, I was able to get pretty close:

```javascript
myDate.toLocaleString('fr-CA', {
  year: 'numeric',
  month: '2-digit',
  day: '2-digit',
  hour: 'numeric',
  minute: '2-digit',
})

// This is the same
myDate.toLocaleString('fr-CA', {
  dateStyle: 'short',
  timeStyle: 'short',
})

// Output: 2022-08-05 13 h 06
```

But the only way to get a separator in there that I found was:

```javascript
const date =  new Date()
const formattedDate = new Date().toLocaleString('fr-CA', {
  dateStyle: 'short',
})
const formattedTime = new Date().toLocaleString('fr-CA', {
  timeStyle: 'short',
})
const formattedDateTime = `${ formattedDate } à ${ formattedTime }`
```

I do have a string translation library with variable insertion that allows me to do the English version of this pretty well, but isn't it unfortunate that I had to do that?

Basically, we have locales that define the "visuals" of the format. The order of the numbers. The separators between the numbers. The separators between hours and minutes. The separators between the date and time.

All of these factors can't be overridden. It's basically just magic that happens behind the scenes that we have no power over. I really, really wish I could pick `fr-CA`'s defaults and _work from there_ instead of having to either take all of its magic or none of it.

Additionally, I'm lucky; I'm from Canada, and the ISO-8601 representation of a date (just the 2022-12-30 part; not the whole machine-friendly string) is what JS understands as a short date for me. But what if someone wants that, but they're not aware of that? The built-in formatter is of no help to them.

**Comment by ptomato** [comment]
Thanks for the detailed use cases! I agree, these would be good use cases for a string format minilanguage.

The point about starting from localized defaults and tweaking a few things is, I fear, difficult to write code for in a general case, so it's probably best to consider it for one or a few locales, like `fr-CA` and `en-CA` in your case.

You can pick `fr-CA`'s defaults and work from there, using [`formatToParts()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/formatToParts), which gives you the same localized string as `toLocaleString()` but broken up into segments that are tagged with the type of data that they represent. It probably gives you _more_ information than you need and so your code using it would be on the verbose side, but could be encapsulated. For example, doing some of the stuff you mentioned:
```js
function formatCustom(plainDateTime) {
  const formatter = new Intl.DateTimeFormat('fr-CA', {
    dateStyle: 'short',
    timeStyle: 'short'
  });

  const { year, month, day, hour, minute } = formatter
    .formatToParts(plainDateTime)
    .reduce((result, { type, value }) => {
      if (type !== 'literal') {
        result[type] = value;
      }
      return result;
    }, {});

  return `${year}-${month}-${day} à <strong>${hour}:${minute}</strong>`;
}
```

It would be easier to do things like this if `formatToParts()` gave more information about the purpose of each separator that it splits out, which is an issue that you could raise at https://github.com/tc39/ecma-402.

I do think that HTML and Markdown are out of scope for a facility that's part of the core JS language. Early JS included special HTML facilities like `"my string".bold()` that wrapped strings in HTML tags but those have been discouraged for a long time. In my opinion, a good formatting API should allow this but not give it special consideration.

**Comment by Ariane-B** [comment]
What if we could do something like this?

```javascript
const  localeDateFormatParams = Intl.getDateFormatParamsForLocale('fr-CA');
console.log(localeDateFormatParams);

// Would output something like:
const localeDateFormatParams = {
  separators: {
    afterDay: '-',
    afterMonth: '-',
    afterDate: ' à ',
  },
  // And much much more info about the minute details of how this locale works
}
```

And then use it like:

```javascript
// An object for a locale would mean "I'm about to define to you the slightest details of a locale that may or may not exist"
const formatter = new Intl.DateTimeFormat(localeDateFormatParams, {
  dateStyle: 'short',
  timeStyle: 'short'
})
```

As for your code example, the existence of `formatToParts` is interesting and I didn't know about it, but wouldn't it be simpler to simply format the date and time separately in `fr-CA` or `en-CA` or a combination of both? Way fewer lines.

**Comment by ptomato** [comment]
It would be fewer lines to use `dateStyle` and `timeStyle` separately if you only need to change the date/time separator string, but I intentionally did it the long way to illustrate what you might do to replace the ` h ` separator with a colon as well.

ECMA-402 have previously indicated that they are skeptical about such changes to Intl.DateTimeFormat: https://github.com/tc39/ecma402/issues/554#issuecomment-796903218 There is a data-driven API being discussed at https://github.com/tc39/ecma402/issues/210 that I think would address this.

**Comment by Ariane-B** [comment]
Oh, I see! Thanks. :)

My sneaky way out of the colon issue was to use `en-CA` for the time no matter what language the page is being displayed in lol

**Comment by leonyu** [comment]
I think the problem with string formatting is the concept and idioms themselves are very English-centric.

For example, most implementation assumes 

* `MMMM` to be `January` and `MMM` to be `Jan`
* `DDDD` to be `Monday` and `DDD` to be `Mon`

However, this does not even translate to other Latin alphabet languages, for example in French, the equivalent of 3-letter month is 4 letters -- `janv.` for `Jan` and `févr.` for `Feb`. Does that mean we need to use `MMMM` for short month in French? And maybe `MMMMM` for long form? ([Ref](https://web.library.yale.edu/cataloging/months)).

Also how does the string format apply to languages where people do not use space separators? (e.g. Asian languages) Does `MMMM` mean two 2-letter `MM`? (e.g. `MM`+`MM`) or does it mean 4-letter `MMMM`?

**Comment by Ariane-B** [comment]
We just need to find the language with the most different ways to express, say, a month, and prepare a scale of complexity based on that. After that, it's probably relatively trivial to just simplify for the sooner languages. The concept of a short month doesn't exist in your language at all? Great! When you ask for a short month, you get the only month that exists.

**Comment by leonyu** [comment]
OK, in that case, let's say the hypothetical "most different" language is some French dialect and their 4-letter word abbreviation for Month is represented by `MMMM`. The 1-3 letter spots (`M`, `MM`, `MMM`) are occupied by different numeric month representations. 

Since as you say just simplify for simpler languages and English is simpler: 

* `M`, `MM`, `MMM` -- numeric month
* `MMMM` -- Jan, Feb, Mar..
* `MMMMM` -- January, February...

Would you be happy with that? I am pretty sure not.

**Comment by Ariane-B** [comment]
I'd be satisfied with that. It may not be the ideal solution, but it'd be a workable solution. 

A workable solution is much better than "let's not have a solution because there is no absolutely perfect one".

Though "MMMM" and such isn't the only model in existence. We could, for example, do templates with keywords that would be a bit more explicit than repeated letters. 

This is just me improvising, but keywords like `${ monthTwoLetters }` would be hard to misunderstand. Or maybe we could take a page out of PHP's book with percent signs and lowercase/uppercase letters. Not explicit and *will* require frequent referring to the docs, but short. 

There are lots of possibilities. We should strive to choose the one(s) that will apply best to the most use cases. But even if there isn't a solution that will perfectly match *every use case*, I think it'll still be a significant win if we can cover most of them.

**Comment by pirate** [comment]
> I think the problem with string formatting is the concept and idioms themselves are very English-centric.

This is however commonly accepted in programming and that's not going to be changed by one datetime library. 

The primary use-case for `strftime` is not producing human-readable localized strings, but rather producing consistent machine-readable strings for things like log files, console output, model fields, filenames, URLs, DSLs, etc., where `ISO-8601` is not suitable and the localization features of `toLocaleString`/`DateTimeFormat` actually hinder consistency.

As wonderful as they are, [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) and [RFC 3339](https://tools.ietf.org/html/rfc3339) do not suit all use-cases, there are reasons to need things like these:

- `date +"%Y-%m-%d_%H-%M-%S_%s"` -> `2024-03-08_12-12-50_1709928770` (for filenames)
- `date '+%Y:%s%5N'` -> `2024:170992885283002` for a year + unix timestamp with specific precision (e.g. for a yearly report file where a human needs to be able to see what year it applies to, but the exact time should be only machine-readable ts format to avoid implying it's a monthly/daily report to the human)
- `date +"%m/%d/%H/%M:%S"` -> `03/08/12/17:09` for assembling a URL where only month, day, and hour are folders but the minute+second entries are a file
- for preparing dates for DSLs (e.g. SQL, JQL, GQL, Excel formulas, config files like INI/YAML, etc. can all store and expect a wide variety of delimiters & formats in real world applications)
- etc. just think of more constraints in real life around special characters, whitespace, implied precision, etc. and you can quickly find situations where machine readable timestamps are vastly easier to format in `strftime`-style than trying to shoehorn `Intl.DateTimeFormat` or manually slice and re-assemble `ISO-8601`.

It doesn't need to be the recommended or primary way of producing date strings, but because it's such an industry standard it would remove the need for external packages just to access this feature. As it stands now, [`strftime` is one of the main reasons why people install external JS libraries](https://stackoverflow.com/search?q=Javascript+strftime) like `moment.js`, `date-fns`, etc.

There's a reason `strftime` is so popular already, and we cannot honestly say that all of its existing use-cases are mis-uses (even if we're trying to push people towards localizing all datetimes by default). I think Python's `datetime` library strikes a great balance here as it provides both `strftime` and `DateTimeFormat`-style localization options.

---

Of course you can technically achieve these examples with some hand-written string slicing and assembling code right now: https://github.com/tc39/proposal-temporal/issues/2501#issuecomment-1986520974 but the whole point of `strftime` and `yy:mm:dd` equivalents are to remove these common, often-repeated, often subtly broken hand-written string formatting functions across the world's codebases and follow the principle of least surprise.

**Comment by leonyu** [comment]
> This is however commonly accepted in programming for developer-facing strings and that's not going to be changed by one datetime library.

String format syntaxes are divergent across languages / libraries even only counting English. Unless someone is willing to do the leg work to go to a standards body getting agreement by at least a dozen different countries or a few different industries, this would pretty much be [xkcd on Standards](https://xkcd.com/927/)

**Comment by pirate** [comment]
I've not found that to be the case, see the links I posted for the spec for each language above above. They almost all use a common set of `strftime` parameters or `TR35`, with only a few variations here and there (e.g. `java`'s `SimpleDateFormat` uses `u` for `Day number of week`).

Since both `strftime` and `TR35` are already well established, we definitely just use one or the other, not create a *new* standard.

**Comment by TimJohns** [comment]
Regardless of whether `strftime/TF35` functionality is ultimately incorporated into Temporal and TC39, is there an existing proposal or candidate reference implementation polyfill elsewhere?

I do have an opinion, and I've already expressed it with emoji in this thread, above, but I also need an interim solution for dealing with a frustrating number of non-standard machine-readable formats when calling or linking to third-party services for reservations. I found a ton of modules, articles, blogs, and code snippets, but if one or more are on a future proposal track, that would be good to know here. I very much expect that many of the other folks coming here and reading this thread will be in the same situation and might be dealing with a bit of "analysis paralysis" trying to choose.

**Comment by ptomato** [comment]
Nothing is on a proposal track at the time of writing. I'm not aware of any such libraries built on top of Temporal.

If I had time, I'd develop one or more lightweight libraries for formatting and parsing using various picture string formats (strftime, TR35, etc), to see what sticks in the ecosystem and to help with adoption of Temporal. But that's realistically not going to happen until Temporal v1 is stable.

So if you feel motivated to do it, go ahead!

**Comment by Rikk** [comment]
How many more decades will we have to wait until ECMAScript allows us to customize DateTime formats without writing complex code or searching for a suitable format among a list of hundreds of country-specific locales? Yes, you are decades behind many other languages, as has been extensively explained above by various people.

> eg. format(Date.now(), 'yyyy-MM-dd HH:mm:ss')

It isn't nice to have to resort to a country-specific locale to obtain the same result:
`new Date(Date.now()).toLocaleString('sv');`
`Temporal.Now.plainDateTimeISO().toLocaleString('sv');`

If we had, let's say, a `toCustomString()` method, eg:
`Temporal.Now.plainDateTimeISO().toCustomString('yyyy-MM-dd HH:mm:ss');`
It would provide a stable, reliable, constant, immutable way for presenting dates and times, instead of relying on a country-specific **political** decision.

**Comment by notpushkin** [comment]
> that means CLDR can never update their locale research without randomly breaking websites across the web.

And that’s, of course, one more reason to have custom formatting in the spec! Stable Formatting solves this for this specific format, but it would be nice to have e.g. `dd.MM.yyyy` without having to rely on `ru` always having this format.

**Comment by ptomato** [comment]
I'll quote myself from earlier in the thread:

> ECMA-402 have previously indicated that they are skeptical about such changes to Intl.DateTimeFormat: [tc39/ecma402#554 (comment)](https://github.com/tc39/ecma402/issues/554#issuecomment-796903218) There is a data-driven API being discussed at [tc39/ecma402#210](https://github.com/tc39/ecma402/issues/210) that I think would address this.

I'm going to ask that we end this line of discussion here and stick to the main topic. You can discuss changes to Intl at the issues I linked above. There's no need to argue why a user-supplied string format would be better; they serve different use cases. The existence of this issue assumes that a user-supplied string format would be a helpful API to have, and the purpose is to discuss what it should look like.
